### DISCUSSION TITLE
[vite] Types for Vite plugin

### DISCUSSION THREAD

**Discussion by angelhdzmultimedia** [DanielXMoore/Civet#1645] [Q&A]
Hi 👋,

can't find the types for vite plugin. I tried setting `types: ["@danielx/civet/vite"]"` as other libs do. Enabling `allowJs` in `tsconfig` doesn't nothing.
Also tried renaming `vite.config.ts` to `vite.config.civet` as a desperate measure.

It's working at runtime. My `App.vue` has a `<script src="./App">`  where `./App` is  `./App.civet` and Vite plugin is successfully transforming `\*.civet` to `\*.ts`.

**Comment by edemaine** [comment]
The types are in `dist/unplugin/vite.d.ts` and they should be automatically part of the `"@danielx/civet/vite"` export. I also haven't had trouble in the past, so it's probably something specific to your setup.

So you're failing to get types for the Vite plugin in VSCode when looking at `vite.config.ts`? Did you try restarting VSCode?

**Reply by angelhdzmultimedia** [reply]
> The types are in `dist/unplugin/vite.d.ts` and they should be automatically part of the `"@danielx/civet/vite"` export. I also haven't had trouble in the past, so it's probably something specific to your setup.
> 
> So you're failing to get types for the Vite plugin in VSCode when looking at `vite.config.ts`? Did you try restarting VSCode?

Thanks. Will investigate further. Of course I did the most mandatory thing and restarted Vue/TypeScript language servers.

K, while typing this message to avoid having to edit it later, I restarted VSCode. Still complaining about the module types.

`tsconfig.json`
```ts
{
  "name": "civet-test",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "scripts": {
    "dev": "vite",
    "build": "vue-tsc -b && vite build",
    "preview": "vite preview"
  },
  "dependencies": {
    "@danielx/civet": "^0.9.1",
    "civet": "link:@types/@danielx/civet",
    "vite": "latest",
    "vue": "^3.5.13"
  },
  "devDependencies": {
    "@vitejs/plugin-vue": "^5.2.1",
    "typescript": "~5.6.2",
    "vue-tsc": "^2.1.10",
    "vite": "link:@types/@danielx/civet/vite"
  }
}

`vite.config.ts`
```ts
import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'
import civetVitePlugin from '@danielx/civet/vite';

// https://vite.dev/config/
export default defineConfig({
  plugins: [
    vue(),
    civetVitePlugin({}),
  ],
})
```


![image](https://github.com/user-attachments/assets/cc9c422e-95b9-4222-95f7-d74d07a890b1)

# Env
Node: 23.3.0
OS: Kubuntu 24.10
VSCode: 1.96.0

Project with: `pnpm create vite` > Vue > TypeScript.

**Reply by edemaine** [reply]
Hmm, I actually see the same on a project of mine. It looks like it's the Civet LSP that's failing to pick up the types. Can you confirm that, if you temporarily disable the Civet VSCode extension and enable the TypeScript VSCode extension, then it works OK?

**Reply by angelhdzmultimedia** [reply]
> Hmm, I actually see the same on a project of mine. It looks like it's the Civet LSP that's failing to pick up the types. Can you confirm that, if you temporarily disable the Civet VSCode extension and enable the TypeScript VSCode extension, then it works OK?

On it as fast as I can!

**Reply by angelhdzmultimedia** [reply]
![image](https://github.com/user-attachments/assets/525b88f5-76d6-43f8-af91-dd4903738fae)

(Reopened VSCode just in case)

**Reply by edemaine** [reply]
OK, it seems to be some kind of nondeterministic behavior then, I guess. Now I can't reproduce it, with both plugins enabled. 😦

**Reply by angelhdzmultimedia** [reply]
I even disabled Vue Official extension (formerly Volar). Strange.

**Reply by angelhdzmultimedia** [reply]
### Update
Yeah apparently Civet extension might be at cause. I made few tests with both the Vue Official and Civet extensions disabled. No error. Then enabled Vue extension, and closed/open VSCode. No errors.
Then enabled Civet extension. Error immediately came back. 


https://github.com/user-attachments/assets/2fb1f720-d312-429d-9012-fd494efa73b0

**Reply by angelhdzmultimedia** [reply]
Tested on all versions down to `0.3.20`.

**Reply by edemaine** [reply]
Thanks for continuing to test! That behavior (only broken with Civet plugin) would suggest that it's an LSP bug, presumably in `resolveModuleNames`.

But it seems like that function should just be calling TypeScript's resolver and successfully resolving. Adding some logs confirms it...

So I'm back to not being sure of the cause. It doesn't help that I can reproduce the problem only sporadically (currently not at all). 😦

**Reply by angelhdzmultimedia** [reply]
No worries, I will keep researching and trying to solve it on my own.
I added the types manually so I can continue with testing Civet with Vue/Vite. 
```ts
 "types": ["/home/angelhdz/Documents/Programming/TypeScript/projects/vue-electron/civet-test/node_modules/.pnpm/@danielx+civet@0.9.1_typescript@5.6.3/node_modules/@danielx/civet/dist/unplugin/vite.d.ts"]
 ```
 The error went away. Then I installed `@vitejs/plugin-vue-jsx` and TSX is working in *.civet files.
 
 The only issue for now is `Could not find a declaration file for module './App.vue'. '/home/angelhdz/Documents/Programming/TypeScript/projects/vue-electron/civet-test/src/App.vue' implicitly has an 'any' type.` when importing a Vue component either in `main.ts` or in a SFC `<script setup lang="ts">
import C from './C'
</script>` so apparently there are still Vue or Vite issues with Civet extension.

But it's rendering :)

![image](https://github.com/user-attachments/assets/3e70041f-94bc-4906-9a01-879b5499be00)

**Comment by nicholaides** [comment]
At a glance it seems like you may be experiencing this bug: https://github.com/DanielXMoore/Civet/issues/1634

**Reply by edemaine** [reply]
I don't think so. These are `exports` of Civet's `package.json`; no `references` in any `tsconfig`s that I know of.

**Reply by angelhdzmultimedia** [reply]
> At a glance it seems like you may be experiencing this bug: #1634

Excellent info! Appreciate it. Subscribing!

**Reply by nicholaides** [reply]
Vite's official project templates use references. Because the civet LSP doesn't understand references, the LSP won't load the tsconfig files and so won't apply the right settings. @angelhdzdev if your root .tsconfig file uses the `reference` key that's going to cause problems for you (even if this particular problem isn't cause by that). #1634 describes a quick workaround.

**Reply by angelhdzmultimedia** [reply]
> At a glance it seems like you may be experiencing this bug: #1634

Your solution fixed all my issues, even the ones I didn't mention, like the one you experienced with DOM API, in my case I couldn't use console.log, it complained about adding `"DOM"` to `lib` but it was already there...
And now my Vue components can be imported without errors (only that I can't use the lang="civet" on the script tag or the .civet extension on the import but I can live with it).

**Reply by angelhdzmultimedia** [reply]
### Update
Guess what guys? 🤡

I was trying Civet @@decorators, and switched from TSX to normal Vue code using defineComponent, was having fun.

Suddenly...

I noticed... i was coding blind. I didn't have intellisense!!! 
Hovering things won't complete the types.

So I full of joy, obviously, quitstarted VSCode...

and...

[HORROR.wav]

ALL. ERRORS. CAME. BACK. 
![image](https://github.com/user-attachments/assets/6573e8a0-d587-48e2-a2a9-7c0a4023d3fb)

**Reply by angelhdzmultimedia** [reply]
Wondering if we can use extend instead of references?  Or Civet also does not understand it? 🤔

**Reply by edemaine** [reply]
Yes, `extends` should work fine.

**Reply by angelhdzmultimedia** [reply]
Confirming. `extends` solved it.


```ts
  "extends": [
    "./src/tsconfig.app.json",
   "./tsconfig.node.json" 
  ],
```

Reloaded VSCode.

No errors.

But wondering what implications will this have for Vite...

Anyways I think this discussion is solved?
